### ISSUE TITLE
__webpack_public_path__ ignored in entry point when hot reloading using webpack-dev-server

### ISSUE THREAD

**Issue by noisecapella** [webpack/webpack#1650]
The webpack [documentation](https://webpack.github.io/docs/configuration.html#output-publicpath) says, if you don't know `publicPath` when compiling you can specify `__webpack_public_path__` in your entry point. `__webpack_public_path__` becomes `__webpack_require__.p` in the webpack bundle. It appears that after hot reloading there are at least two copies of the function `__webpack_require__` since the `p` property is only set on one of them. My hot reload is failing because whatever copy of `__webpack_require__` it's using to hot reload doesn't have `p` set to the new value. If I set `publicPath` in `webpack.config.js` the reload works fine.

Does this sound correct? Thanks for any guidance!

**Comment by sokra** [comment]
yep, that's true.

**Comment by p0wl** [comment]
also ran into this. really hard to find the cause... any idea for a solution besides setting the public path in the config?

The problem is, that with hmr enabled, the variable `__webpack_require__` is created for each module (id): https://github.com/webpack/webpack/blob/master/lib/HotModuleReplacement.runtime.js#L30-L34

because of the highlighted lines, the public path is copied at module load and no longer connected to the "global" __webpack_require__. So changing the public path (`__webpack_public_path__`) will only change the public path for the `__webpack_require__` that was created for this module.

tl;dr: look at the generated output of a module:
```
/***/ 53:
/***/ function(module, exports, __webpack_require__) {
    __webpack_require__.p = 'https://mycdn.com/myPublicPath/';
/***/ },
```

if HMR is not enabled, `__webpack_require__` is the global instance, where the public path (`.p`) is read from.
if HMR is enabled, each module gets a specialized `__webpack_require__` injected, which is no longer pointing to the global `__webpack_require__` and therefore changes at this object are not visible to other modules or the webpack runtime.

**Comment by epsitec** [comment]
I don't see any workaround for this issue?

**Comment by sokra** [comment]
It's actually easy to fix...

**Comment by p0wl** [comment]
:+1:
